### PR TITLE
Apply Beman Standard and add more examples/docs

### PR DESCRIPTION
**Pull request by neatudarius**
- [x] Apply Beman Standard: CMAKE.SKIP_EXAMPLES + CMAKE.SKIP_TESTS
- [x] Add examples
- [x] Add docs: Readme
- [x] Add docs: Examples
- [x] Add Compiler Explorer links

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -42,8 +45,8 @@ configure_file(
   "${PROJECT_SOURCE_DIR}/include/beman/iterator_interface/config.hpp.in"
   "${PROJECT_BINARY_DIR}/include/beman/iterator_interface/config.hpp" @ONLY)
 
-# Build the tests if enabled via the option ITERATOR_INTERFACE_ENABLE_TESTING
-if(ITERATOR_INTERFACE_ENABLE_TESTING)
+# Build the tests if enabled via the option BEMAN_ITERATOR_INTERFACE_BUILD_TESTING
```
Comment (camio) [inline CMakeLists.txt:48]: I think the code is self-explanatory without the comment.
```suggestion
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -79,11 +82,16 @@ target_include_directories(
 add_subdirectory(src/beman/iterator_interface)
 add_subdirectory(include/beman/iterator_interface)
 
-add_subdirectory(examples)
-if(ITERATOR_INTERFACE_ENABLE_TESTING)
+# [CMAKE.SKIP_TESTS]
```
Comment (camio) [inline CMakeLists.txt:85]:
```suggestion
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -79,11 +82,16 @@ target_include_directories(
...
+if(BEMAN_ITERATOR_INTERFACE_BUILD_TESTING)
   add_subdirectory(tests/beman/iterator_interface)
 endif()
 
+# [CMAKE.SKIP_EXAMPLES]
```
Comment (camio) [inline CMakeLists.txt:90]:
```suggestion
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -27,8 +27,11 @@ option(BEMAN_ITERATOR_INTERFACE_USE_DEDUCING_THIS
   "Make use of C++23 \"deducing this\" feature (P0847R7). Turn this off for non-conforming compilers."
   ${COMPILER_SUPPORTS_DEDUCING_THIS})
 
-option(ITERATOR_INTERFACE_ENABLE_TESTING
-       "Enable building tests and test infrastructure" ${PROJECT_IS_TOP_LEVEL})
+option(BEMAN_ITERATOR_INTERFACE_BUILD_TESTING
```
Comment (camio) [inline CMakeLists.txt:30]: Should we call this `BEMAN_ITERATOR_INTERFACE_BUILD_TESTS` for consistency with `BEMAN_ITERATOR_INTERFACE_BUILD_EXAMPLES`?

**Comment by neatudarius** [comment]
@steve-downey @camio , can you take a look?

File: CMakeLists.txt (hunk 4 above)
Comment (camio) [reply CMakeLists.txt:30]: Yeah, I suppose so if you agree.

Comment (neatudarius) [reply CMakeLists.txt]: done

**Review by steve-downey — APPROVED** [review]
👍

File: CMakeLists.txt [hunk 5]
```diff
@@ -79,11 +81,14 @@ target_include_directories(
 add_subdirectory(src/beman/iterator_interface)
 add_subdirectory(include/beman/iterator_interface)
 
-add_subdirectory(examples)
-if(ITERATOR_INTERFACE_ENABLE_TESTING)
+if(BEMAN_ITERATOR_INTERFACE_BUILD_TESTS)
```
Comment (camio) [inline CMakeLists.txt:84]:
```suggestion
if(BEMAN_ITERATOR_INTERFACE_BUILD_TESTS)
  enable_testing()
```

File: README.md [hunk 6]
```diff
@@ -18,53 +17,379 @@ Documentation and associated papers are licensed with the Creative Commons Attri
 
 // SPDX-License-Identifier: CC-BY-4.0
 
-The intent is that the source and documentation are available for use by people implementing their own optional types as well as people using the optional presented here as-is.
+The intent is that the source and documentation are available for use by people implementing their iterator types as well as people using the iterator_interface presented here as-is.
```
Comment (camio) [inline README.md:20]:
```suggestion
The intent is that the source and documentation are available for use by people implementing their iterator types.
```

File: examples/filter_int_iterator.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    // Default constructor creates an end-of-range iterator.
+    filtered_int_iterator() : m_it_begin(nullptr) {}
+
+    // Constructor for the beginning of the sequence.
+    filtered_int_iterator(int* it_begin, int* it_end, Pred pred) : m_it_begin(it_begin), m_it_end(it_end), m_pred(std::move(pred)) {
+        m_it_begin = std::find_if(m_it_begin, m_it_end, m_pred);
+    }
+
+    // A forward iterator based on iterator_interface usually requires
+    // three user-defined operations.  since we are adapting an existing
```
Comment (RaduNichita) [inline examples/filter_int_iterator.cpp:28]:
```suggestion
    // three user-defined operations.  Since we are adapting an existing
```

Comment (RaduNichita) [inline examples/filter_int_iterator.cpp:28]:
```suggestion
    // three user-defined operations.  since we are adapting an existing
```

maybe mentions which operations?

File: examples/filter_int_iterator.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
+// examples/filter_int_iterator.cpp -*-C++-*-
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// [P2727](https://wg21.link/P2727) example:
+// An iterator that allows filtering int elements of a sequence.
```
Comment (RaduNichita) [inline examples/filter_int_iterator.cpp:5]: I would leave an empty line between the comments and the include part

File: examples/README.md [hunk 9]
```diff
@@ -1,22 +1,29 @@
...
+* [P2727](https://wg21.link/P2727): local [./filter_int_iterator.cpp](./filter_int_iterator.cpp) or [filter_int_iterator.cpp@Compiler Explorer](https://godbolt.org/z/q6933enqe)
 
-Check [sample](sample.cpp) for basic `beman.iterator_interfaces` library usage.
+## Local Build and Run
 
-Build and run instructions:
 ```shell
 # build
-$ cmake --workflow --preset gcc-14
+$ cmake --workflow --preset gcc-14 # or choose one from --list-presets
```
Comment (RaduNichita) [inline examples/README.md:20]:
```suggestion
$ cmake --workflow --preset gcc-14 # or choose one from --list-presets
```

update the command to use `gcc-debug` or `gcc-release`

File: examples/README.md [hunk 10]
```diff
@@ -1,22 +1,29 @@
...
+
+# run repeated_chars_iterator.cpp
+$ .build/gcc-14/examples/Asan/beman.iterator_interface.examples.repeated_chars_iterator
```
Comment (RaduNichita) [inline examples/README.md:23]: update the path based on the previous comment

File: examples/repeated_chars_iterator.cpp [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    difference_type m_pos;
+};
+
+int main() {
+    // Create a repeated_chars_iterator that iterates over the sequence "foo" repeated indefinitely:
+    // "foofoofoofoofoofoo...". Will actually extract a prefix of the sequence and insert it into a std::string.
+    constexpr const std::string_view target = "foo";
+    constexpr const auto len = 7;                                        // Number of extracted characters from the sequence.
+
+    // Create iterators that iterate over the sequence "foofoofoofoofoofoo...".
```
Comment (RaduNichita) [inline examples/repeated_chars_iterator.cpp:53]:
```suggestion
    // Create iterators that go over the sequence "foofoofoofoofoofoo...".
```

File: README.md [hunk 12]
```diff
@@ -18,53 +17,379 @@ Documentation and associated papers are licensed with the Creative Commons Attri
...
+// Create a filtered_int_iterator that iterates over the sequence {1, 2, 3, 4, 10, 11, 101, 200, 0},
+// skipping odd numbers. 0 is not skipped, so it will be the last element in the sequence.
+std::array a = {1, 2, 3, 4, 10, 11, 101, 200, 0};
+filtered_int_iterator it{std::begin(a), std::end(a), [](int i) { return i % 2 == 0; }};
+
+while (*it) {
+    std::cout << *it << " ";
+    ++it;
+}
+std::cout << "\n";
```
Comment (RaduNichita) [inline README.md:64]: put here 

```
Expected output at STDOUT
```

like in the other example

File: README.md [hunk 13]
```diff
@@ -18,53 +17,379 @@ Documentation and associated papers are licensed with the Creative Commons Attri
...
+```
+
 ## How to Build
 
+### Compiler Support
+
+This is a modern C++ project which can be compiled with the latest C++ standards (**C++20 or later**).
```
Comment (RaduNichita) [inline README.md:71]:
```suggestion
This is a modern C++ project which can be compiled with the latest C++ standards (**C++20 or later**).
```

I would rephrase this to avoid the word "modern". I would say something like "This project is recommended to be run with a compiler having support for C++20 or later.

File: README.md [hunk 14]
```diff
@@ -18,53 +17,379 @@ Documentation and associated papers are licensed with the Creative Commons Attri
...
+-- Generating done (0.0s)
+-- Build files have been written to: /home/dariusn/git/Beman/iterator_interface/.build/gcc-debug
+
+Executing workflow step 2 of 3: build preset "gcc-debug"
+
+[15/15] Linking CXX executable tests/beman/iterator_interface/beman.iterator_interface.tests
+
+Executing workflow step 3 of 3: test preset "gcc-debug"
+
+Test project /home/dariusn/git/Beman/iterator_interface/.build/gcc-debug
```
Comment (RaduNichita) [inline README.md:175]:
```suggestion
Test project /home/dariusn/git/Beman/iterator_interface/.build/gcc-debug
```

add something generic here

File: README.md [hunk 15]
```diff
@@ -18,53 +17,379 @@ Documentation and associated papers are licensed with the Creative Commons Attri
...
+-- Generating done (0.0s)
+-- Build files have been written to: /home/dariusn/git/Beman/iterator_interface/.build/gcc-release
+
+Executing workflow step 2 of 3: build preset "gcc-release"
+
+[15/15] Linking CXX executable tests/beman/iterator_interface/beman.iterator_interface.tests
+
+Executing workflow step 3 of 3: test preset "gcc-release"
+
+Test project /home/dariusn/git/Beman/iterator_interface/.build/gcc-release
```
Comment (RaduNichita) [inline README.md:235]:
```suggestion
Test project /home/dariusn/git/Beman/iterator_interface/.build/gcc-release
```

Add something generic here

**Review by RaduNichita — COMMENTED** [review]
LGTM

File: README.md (hunk 13 above)
Comment (neatudarius) [reply README.md:71]: It's the same from other repos. And btw, I think modern fits well here. Let's open a topic for "standard" description for repos in Beman.

**Comment by neatudarius** [comment]
Merging this for Meeting C++ presentation. I will do follow-up PRs. Please continue to review!

**Comment by RaduNichita** [comment]
LGTM! 🎉

**Comment by steve-downey** [comment]
"Contemporary" has been suggested as a replacement for "modern", since modern is a decade old style?
